### PR TITLE
[NNAPI EP] Update NnapiExecutionProvider::GetCapability() to use partitioning utils

### PR DESCRIPTION
**Pull request by edgchen1**
**Description**
Update NnapiExecutionProvider::GetCapability() to use partitioning utils.

**Motivation and Context**
Re-use updated EP partitioning logic.

### CODE REVIEW

File: onnxruntime/core/providers/nnapi/nnapi_builtin/nnapi_execution_provider.cc [hunk 1]
```diff
@@ -1,28 +1,36 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
-#include "nnapi_execution_provider.h"
+#include "core/providers/nnapi/nnapi_builtin/nnapi_execution_provider.h"
```
Comment (YUNQIUGUO) [inline onnxruntime/core/providers/nnapi/nnapi_builtin/nnapi_execution_provider.cc:4]: nit: Can these header files be in order?

File: onnxruntime/core/providers/partitioning_utils.h [hunk 2]
```diff
@@ -18,27 +18,73 @@ class Node;
...
 namespace utils {
 
+/**
+Called to check whether a node is supported.
+
+@param node The node to check.
+
+@return Whether the node is supported.
+*/
+using IsNodeSupportedFn = std::function<bool(const Node& node)>;
```
Comment (YUNQIUGUO) [inline onnxruntime/core/providers/partitioning_utils.h:28]: what does all the `Fn` in the name stand for?

File: onnxruntime/core/providers/nnapi/nnapi_builtin/nnapi_execution_provider.cc (hunk 1 above)
Comment (edgchen1) [reply onnxruntime/core/providers/nnapi/nnapi_builtin/nnapi_execution_provider.cc:4]: I believe they are :)
https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

File: onnxruntime/core/providers/partitioning_utils.h (hunk 2 above)
Comment (edgchen1) [reply onnxruntime/core/providers/partitioning_utils.h:28]: function

File: onnxruntime/core/providers/partitioning_utils.cc [hunk 3]
```diff
@@ -3,97 +3,173 @@
...
+
+  std::vector<std::vector<const Node*>> supported_groups{};
+
+  // number of inputs from unprocessed nodes (in-degree) per node
+  std::unordered_map<NodeIndex, size_t> in_degree{};
+  // nodes that are ready to process
+  std::deque<const Node*> nodes_to_process{};
+  // nodes that will be processed when considering the next partition node group
+  std::deque<const Node*> nodes_to_process_with_next_group{};
+
```
Comment (skottmckay) [inline onnxruntime/core/providers/partitioning_utils.cc:73]: Not clear to me what the 'next partition node group' means. We create a partition from supported nodes and essentially ignore all the others, so conceptually I'm not sure there's an 'unsupported partition'.

File: onnxruntime/core/providers/partitioning_utils.cc [hunk 4]
```diff
@@ -3,97 +3,173 @@
...
-      current = supported_to_visit.front();
-      supported_to_visit.pop();
-    } else {
-      current = unsupported_to_visit.front();
-      unsupported_to_visit.pop();
+    const Node& node = *nodes_to_process.front();
+    nodes_to_process.pop_front();
+
+    const bool is_node_supported = is_node_supported_fn(node);
+
```
Comment (skottmckay) [inline onnxruntime/core/providers/partitioning_utils.cc:125]: Do we still always start with unsupported nodes? It's important to do so as the unsupported nodes may produce outputs that enable supported nodes to form a larger partition.

File: onnxruntime/core/providers/partitioning_utils.cc (hunk 3 above)
Comment (edgchen1) [reply onnxruntime/core/providers/partitioning_utils.cc:73]: yes, we only put supported nodes in partitions. in some cases, we will have more than one partition. we form the partitions one at a time so the next partition is just the subsequent one, if any.

File: onnxruntime/core/providers/partitioning_utils.cc (hunk 4 above)
Comment (edgchen1) [reply onnxruntime/core/providers/partitioning_utils.cc:125]: now, we don't necessarily start with unsupported nodes. instead, for a single partition, we process:
- supported nodes - these are included in the current partition
- unsupported nodes that are not downstream of the current partition. the other unsupported nodes can't be processed yet as they form the boundary of the current partition

this also allows the unsupported nodes to be processed and enable other potentially supported nodes to become ready to process.

File: onnxruntime/core/providers/nnapi/nnapi_builtin/nnapi_execution_provider.cc (hunk 1 above)
Comment (YUNQIUGUO) [reply onnxruntime/core/providers/nnapi/nnapi_builtin/nnapi_execution_provider.cc:4]: Not so familiar with the convention before.. sorry about that. Thanks for the link!!

File: onnxruntime/core/providers/partitioning_utils.cc (hunk 4 above)
Comment (skottmckay) [reply onnxruntime/core/providers/partitioning_utils.cc:125]: So it's like a wave that flows through the model, starting at the root node/s, that stops when it gets to a point where the only remaining nodes at the edge of the wave are unsupported nodes that are outputs from supported nodes? As the wave goes through it flows over both unsupported and supported nodes, adding the supported nodes to the current partition.

Following that, the next wave starts with the nodes where the previous wave stopped.

**Review by skottmckay — DISMISSED** [review]
:shipit:

File: onnxruntime/core/providers/partitioning_utils.cc (hunk 4 above)
Comment (gwang-msft) [reply onnxruntime/core/providers/partitioning_utils.cc:125]: Seems the logic has changed a lot, maybe we can use a bit more comments here for futrue reader of these code,
For example, check if all the inputs are available is now part of the checking if the node's indegree == 0, ...

Comment (edgchen1) [reply onnxruntime/core/providers/partitioning_utils.cc:125]: @skottmckay that's a good way of describing it. I'll add some comments

File: onnxruntime/core/providers/partitioning_utils.cc [hunk 5]
```diff
@@ -51,6 +51,38 @@ std::string NodeGroupDebugString(const Container& group, bool show_all = false)
...
+Conceptually, we consider nodes in a sequence of waves starting from the root nodes. One wave produces at most one
+group. A wave flows over nodes in topological order, adding supported nodes to the current group, and stops at the
+border of the current group. The next wave starts where the previous wave stopped.
+
+In the topological sort, we maintain a set of nodes that have no unprocessed inputs from which we select the next node
+to process.
+
+When selecting the next node to process, we first take:
+- a supported node (which will be part of the group)
+- an unsupported node that is not an output of any node in the group
```
Comment (gwang-msft) [inline onnxruntime/core/providers/partitioning_utils.cc:74]: nit,  an unsupported node that does not consume an output of any node in the group

**Review by gwang-msft — DISMISSED** [review]

**Review by gwang-msft — APPROVED** [review]

**Pull request merged**